controllers/watcherapi_controller.go

```diff
@@ -18,39 +18,210 @@ package controllers
 
+	// We reached the end of the Reconcile, update the Ready condition based on
+	// the sub conditions
+	if instance.Status.Conditions.AllSubConditionIsTrue() {
+		instance.Status.Conditions.MarkTrue(
+			condition.ReadyCondition, condition.ReadyMessage)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// generateServiceConfigs - create Secret which holds the service configuration
+// NOTE - jgilaber this function is WIP, currently implements a fraction of its
+// functionality and will be expanded of further iteration to actually generate
+// the service configs
+func (r *WatcherAPIReconciler) generateServiceConfigs(
+	ctx context.Context, instance *watcherv1beta1.WatcherAPI,
+	secret corev1.Secret,
+	helper *helper.Helper, envVars *map[string]env.Setter,
+) error {
+	Log := r.GetLogger(ctx)
+	Log.Info("generateServiceConfigs - reconciling")
+
+	db, err := mariadbv1.GetDatabaseByNameAndAccount(ctx, helper, watcher.DatabaseCRName, instance.Spec.DatabaseAccount, instance.Namespace)
+	if err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			condition.ServiceConfigReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			condition.ServiceConfigReadyErrorMessage,
+			err.Error()))
+		return err
+	}
+
+	// replace by actual usage in future iterations
+	_ = db
+	_ = secret
+	_ = envVars
+
+	return nil
+}
+
+func (r *WatcherAPIReconciler) reconcileDelete(ctx context.Context, instance *watcherv1beta1.WatcherAPI, helper *helper.Helper) (ctrl.Result, error) {
+	Log := r.GetLogger(ctx)
+	Log.Info(fmt.Sprintf("Reconcile Service '%s' delete started", instance.Name))
+
+	controllerutil.RemoveFinalizer(instance, helper.GetFinalizer())
+	Log.Info(fmt.Sprintf("Reconciled Service '%s' delete successfully", instance.Name))
 	return ctrl.Result{}, nil
 }
 
+func (r *WatcherAPIReconciler) initStatus(instance *watcherv1beta1.WatcherAPI) error {
+
+	cl := condition.CreateList(
+		// Mark ReadyCondition as Unknown from the beginning, because the
+		// Reconcile function is in progress. If this condition is not marked
+		// as True and is still in the "Unknown" state, we `Mirror(` the actual
+		// failure/in-progress operation
+		condition.UnknownCondition(condition.ReadyCondition, condition.InitReason, condition.ReadyInitMessage),
+		condition.UnknownCondition(condition.InputReadyCondition, condition.InitReason, condition.InputReadyInitMessage),
+		condition.UnknownCondition(condition.ServiceConfigReadyCondition, condition.InitReason, condition.ServiceConfigReadyMessage),
+	)
+
+	instance.Status.Conditions.Init(&cl)
+
+	// Update the lastObserved generation before evaluating conditions
+	instance.Status.ObservedGeneration = instance.Generation
+
+	return nil
+}
+
 // SetupWithManager sets up the controller with the Manager.
 func (r *WatcherAPIReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
```

Comment:
After this patch, the watcherapi controller needs to monitor also secrets. See:

https://github.com/openstack-k8s-operators/barbican-operator/blob/main/controllers/barbicanapi_controller.go#L891

and we should implement something like that to specify the secrets to watch:

https://github.com/openstack-k8s-operators/barbican-operator/blob/main/controllers/barbicanapi_controller.go#L894-L898

Reply:
looking at your links, I see that barbicanapi both `Watches` and `Owns` secrets, do we need both?

Reply:
Yes, that's my understanding:

// Owns defines types of Objects being *generated* by the ControllerManagedBy, and configures the ControllerManagedBy to respond to create / delete / update events by *reconciling the owner object*.

and:

// Watches defines the type of Object to watch, and configures the ControllerManagedBy to respond to create / delete / update events by *reconciling the object* with the given EventHandler.

Reading those descriptions, at this point, I think the controller is only accessing secrets so using only a Watches may be enough but we will need to create new secrets later, so I guess we can add the Owns too.

Reference: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/builder#ControllerManagedBy

Reply:
I see, thanks for the clarification!

Reply:
updated the PR with the change